libraries/botbuilder-core/src/transcriptLogger.ts

```diff
@@ -92,7 +92,22 @@ export class TranscriptLoggerMiddleware implements Middleware {
         while (this.transcript.length > 0) {
             try {
                 const activity: Activity = this.transcript.shift();
-                this.logger.logActivity(activity);
+                // If the implementation of this.logger.logActivity() is asynchronous, we don't
+                // await it as to not block processing of activities.
+                // Because TranscriptLogger.logActivity() returns void or Promise<void>, we capture
+                // the result and see if it is a Promise.
+                const logActivityResult = this.logger.logActivity(activity);
+
+                // If this.logger.logActivity() returns a Promise, a catch is added in case there
+                // is no innate error handling in the method. This catch prevents
+                // UnhandledPromiseRejectionWarnings from being thrown and prints the error to the
+                // console.
+                if (logActivityResult instanceof Promise) {
+                    logActivityResult.catch(err => {
+                        // tslint:disable-next-line:no-console
+                        console.error('TranscriptLoggerMiddleware logActivity failed', err);
+                    });
+                }
             } catch (err) {
```

Comment:
This is something of a nit.

If this err object is of type Error (which it _should_ be but isn't necessarily) then this line will do a toString which will write only a subset of the information being carried to the console.

Wouldn't it therefore be more correct to explicitly check if it was of type Error and if it was then print out more of the fields, otherwise just toString()? (Incidentally, if I remember correctly I don't think JSON.stringify will work on Error.)

In this particular case I'm not sure this matters to much because its just the console. But something to consider in general.

Reply:
How about this?

```ts
try {
  const activity: Activity = this.transcript.shift();
  // Comments comments comments...
  const logActivityResult = this.logger.logActivity(activity);
  // Comments comments comments...

  if (logActivityResult.catch(err => {
    this.transcriptLoggerErrorHandler(err);
  });
} catch (err) {
  this.transcriptLoggerErrorHandler(err);
}

private transcriptLoggerErrorHandler(err: Error|Any): void {
  // tslint:disable:no-console
  if (err instanceof Error) {
    console.error(`TranscriptLoggerMiddleware logActivity failed: "${ err.message }"`);
    console.error(err.stack);
  } else {
    console.error(`TranscriptLoggerMiddleware logActivity failed: "${ JSON.stringify(err) }"`);
  }
  // tslint:enable:no-console
}
```